README.rst
==============================
Вычисление числа 𝛑 по разрядам
==============================

Задание
=======

Реализовать веб-приложение для вывода n-ой цифры в числе Пи

1. На основной странице отображаем таблицу с результатами всех вычислений и
   элемент для добавления новой задачи по получению n-ой цифры.
2. Задачи выполняем асинхронно вне зависимости от реализации.
3. В таблице с результатами отображаем статус задачи (завершена, в процессе,
   ошибка).
4. Обновление статуса задачи на странице также должно быть асинхронное.
5. Каждая строка таблицы должна содержать ссылку на страницу, на которой
   должно выводится число Пи с точностью до n знаков.


Запуск
======

.. code-block:: sh

   $ aria2c -j 5 -x 5 http://buildbot.pypy.org/nightly/py3.7/pypy-c-jit-100194-4b803dfae562-linux64.tar.bz2
   $ tar -xvf pypy-c-jit-100194-4b803dfae562-linux64.tar.bz2
   $ pypy-c-jit-100194-4b803dfae562-linux64/bin/pypy3 -m venv venv3.7
   $ . venv3.7/bin/activate
   $ pip install -U pip wheel setuptools poetry
   $ poetry install
   $ npm install
   $ npm run build
   $ python server.py

Открыть `веб-приложение <http://localhost:50594/>`_.


Реализация
==========

Концепция
---------

Вычисление числа ПИ - это дорогостоящая вычислительная операция. Обслуживание http-запросов - это дорогостоящая операция ввода-вывода.
Обслуживаем трафик асинхронно, вычисляем ПИ на доступных процессорах (cpu).
Такая система с многими обработчиками одного "бизнес-процесса" является источником гонки данных. Точкой синхронизации будет СУБД и её механизм ограничений (``sqlalchemy.UniqueConstraint``).


Воплощение
----------

Главный процесс выделяет следующие ресурсы:

- событийный цикл для обработки http-запросов (``eventloop``)
- очередь задач, которая содержит количество знаком после запятой для расчета цисла ПИ (``task_queue: multiprocessing.Queue``)
- событие-сигнал об изменении в хранилище данных (``database_updated: multiprocessing.Event``)
- процессы в количестве ``multiprocessing.cpu_count`` для запуска вычислений (``workers: multiprocessing.Process``)
- менеджер соединения к СУБД sqlite (``Session: sqlalchemy.<?>``)

В обычном режиме работы sqlite блокирует запись в таблицу от нескольких процессов. С помощью включения журнала ``PRAGMA journal_mode=WAL`` можно избежать исключений о заблокированной БД.


Штучки
------

Для человека "рассчитать очередную точность числа ПИ" - это одна операция. Для системы - это каскад посредников и сред передачи данных.
Я выделил отдельную таблицу в БД для отслеживания этапов обработки:

- ``registred`` - обработчик http-запроса принял данные, зафиксировал в БД
- ``scheduled`` - обработчик http-запроса отправил в очередь на обработку исполнителям в отдельные процессы
- ``taken`` - исполнитель в отдельном процессе забрал из очереди задачу и приступил к вычислениям
-  ``calculated`` - процесс вычисления завершился
- ``stored``- исполнитель зафиксировал результат в источники истины - БД

Каждый из этапов потенциально может завершится с ошибкой, поэтому выделено место под ошибку.

Отображение прогресса вычисления построено с использование "исчезающего" шаблонизатора `Monkberry <https://monkberry.js.org/>`_ на стороне браузера и websockets в качестве транспорта.


Ограничения
-----------

Веб-приложение запускается в единственном экземпляре и эксклюзивно использует своё хранилище данных.

Для меня демо-приложение (тестовое задание изначально) не должно быть прям на 100% работоспособным и выверенным.
Я специально не стал писать код перерасчата (перезапуск задачи) в случае падения. Даже падения, как мне кажется были, бы на на руку - они бы продемонстрировали работу статусов.
Однако на работе я по обыкновению (только если это подразумевает "бизнес-логика") добавляю функцию, которая при старте закидывает в очередь недоделанные задачи с прошлого падения-остановки.


Журнал оживления проекта
========================

Я время от времени возвращаюсь к проекту и обновляю его. Тема с асинхронностью, многозадачностью в питоне и в навыках программиста сейчас одна из актуальных.
Что это все так спохватись сейчас? Асинхронный сервер с процессами я писал в 2010-2011 годах с подобной моделью. Это было просто будничное время за 1000$ в месяц (30000 рублей) без всяких регалий типа сеньор, архитектор, тимлид, техлид, 5-7-9 лет опыта.


Автоматизация сборки
--------------------

Я готовился на вакансию наставника в `HTMLAcademy <https://htmlacademy.ru/>`_ или выполнял задания по программе обучения-трудоустройства `FSD Front-end education <https://www.fullstack-development.com/front-end-education>`_.
Решил: хватит собирать js руками! Пора прикрутить webpack.


Нештатная ситуация (1)
----------------------

Я добавил блок ``finally``, в котором организовал остановку дочерних процессов.
Для сигнала об останове я выбрал ``multiprocessing.Queue.close``::

  Indicate that no more data will be put on this queue by the current process. The background thread will quit once it has flushed all buffered data to the pipe. This is called automatically when the queue is garbage collected.

Я искренне рассчитывал, что метод данный метод просигналит на другой конец о том, что никаких больше задач не будет.
Так вот никакого "сообщения" на другом конце очереди не было получено. Процессы продолжали висеть в ожидании ``multiprocessing.Queue.get``.
Возможно где-то сидит блокировка и как-то нужно особым образом быть обходительным с Queue. На такие мысли наталкивает история `The tragic tale of the deadlocking Python queue <https://codewithoutrules.com/2017/08/16/concurrency-python/>`_.
Я не стал углубляться в такие подробности, потому что не ставил перед собой целью отладки интерпретатора во имя добра и справедливости.
``terminate`` расшевелил процессы.


Нештатная ситуация (2)
----------------------

Обновил sqlalchemy с 1.2 до 1.3 - перестали работать выборки.
Список вычислений приходит, но работа всегда в состоянии ``registered``. Из-за этого создается впечатление, что ничего не работает.
Я провели данные - там все есть. Попросил sqlalchemy показать запрос - он такой, каким я ожидал увидеть.
Я могу писать запросы SQL, я могу взять на вооружение многообразные ORM (ActiveRecord). Но это всегда в рамках задачи. Вот и сейчас пришлось ещё чуть чуть глубже капнуть в SQL, потому что моих навыков не хватило.

Оригинальный запрос состоял из объединения таблицы с количеством знаков после запятой с прогрессом вычисления. Нас интересует информация о знаку - используем группировку по знаку. Нас интересует актуальный статус вычисления - используем сортировку:

.. code-block:: py

    results = (session
        .query(DigitNumber.digit_number, Event.progress, Event.result)
        .join(Event)
        .group_by(DigitNumber)
        .order_by(-DigitNumber.digit_number, -Event.progress))

.. code-block:: sql

   SELECT digit_number.digit_number AS digit_number_digit_number, event.progress AS event_progress, event.result AS event_result 
   FROM digit_number JOIN event ON digit_number.digit_number = event.digit_number GROUP BY digit_number.digit_number ORDER BY -digit_number.digit_number, -event.progress


Исправление заключалось в использовании ``.having(Event.progress == func.MAX(Event.progress))``, что б наверняка движок вернул "самый актуальный статус". Результат:

.. code-block:: sql

   SELECT digit_number.digit_number AS digit_number_digit_number, event.progress AS event_progress, event.result AS event_result 
   FROM digit_number JOIN event ON digit_number.digit_number = event.digit_number GROUP BY digit_number.digit_number 
   HAVING event.progress = max(event.progress) ORDER BY -digit_number.digit_number


Технический долг
================

Всега остается что-то на потом. Всегда есть вариант хакнуть и вариант разобраться и подобрать верное решение (hacker vs developer). Вот мой список долга перед разработкой.


Где же висят воркеры, когда порт занят и без задач нужно выходить
-----------------------------------------------------------------

См "Нештатная ситуация (1)". В статье есть команды подключения к процессу дебаггером.
Возможно "лучшего" способа и не будет. За то появится глубокое понимания внутреннего устройства и навых отладки питона.


Логирование
-----------

Давняя моя больная мозоль. Для меня было проблемой вообще включить вывод "access logs" для aiohttp. Я понимаю иерархию, мотивацию принторов-обработчиков, иерархии уровней. Но как это заставить работать - нет.
Я до сих пор искренне изумляюсь по какой причине нужно в самом начале забабахать ``logging.basicConfig(level=logging.DEBUG)``, а потом оно уже не будет иметь эффекта.

С логирование связана ещё одна проблема. Система исправно работала, но были проблемы в отображении - выборке данных. См "Нештатная ситуация (2)". По хорошему нужно расставить логи для отображения жизнедеятельности воркеров.


Тестирование
------------

Тестирование на работе не является для меня проблемой. Я делаю их по своей инициативе.
Даже так: я не считаю, что я выполнил работу, если не поставил тестов.
Я предодевал разные трудности на пусти создания тестового окружения. Типа "корутина создана не в этом потоке, корутина создана не в этом эвентлупе". Но для демо проекта у меня пока не хватает мотивации включать тесты.


Переподключение к websocket
---------------------------

Не знаю отнести ли проблему отсутствия переподключения к ограничению или к долгу.



HISTORY.rst
==============================
Вычисление числа 𝛑 по разрядам
==============================


Задание
=======

Реализовать веб-приложение для вывода n-ой цифры в числе Пи

1. На основной странице отображаем таблицу с результатами всех вычислений и
   элемент для добавления новой задачи по получению n-ой цифры.
2. Задачи выполняем асинхронно вне зависимости от реализации.
3. В таблице с результатами отображаем статус задачи (завершена, в процессе,
   ошибка).
4. Обновление статуса задачи на странице также должно быть асинхронное.
5. Каждая строка таблицы должна содержать ссылку на страницу, на которой
   должно выводится число Пи с точностью до n знаков.


Запуск
======

.. code-block:: sh

   $ aria2c -j 5 -x 5 http://buildbot.pypy.org/nightly/py3.6/pypy-c-jit-97722-6e344fc026c3-linux64.tar.bz2
   $ tar -xvf pypy-c-jit-97722-6e344fc026c3-linux64.tar.bz2
   $ pypy-c-jit-97722-6e344fc026c3-linux64/bin/pypy3 -m venv venv
   $ . venv/bin/activate
   $ pip install -U pip wheel setuptools pipenv
   $ pipenv sync
   $ yarn install
   $ npm run build
   $ python server.py

Открыть `веб-приложение <http://127.0.0.1:8080/>`_.


Ход решения
===========

Много контента остается по ту сторону вашего экрана, то есть здесь со мной на рабочем месте и во время выполнения. Не
хочу, что бы добро пропадало зря. Плюс, зафиксировав мысль или решение мне легче двигаться дальше. Читатель увидит "куда
уходит время" в истории коммитов.

Выбраны инструменты:

 - интерпретатор `pypy3.5 <http://buildbot.pypy.org/nightly/py3.5/pypy-c-jit-95404-ce637db93b36-linux64.tar.bz2>`_
 - фреймворк `aiohttp <https://docs.aiohttp.org/en/stable/>`_
 - канал взаимодействия `WebSockets <https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API>`_
 - очередь задач `multiprocessing.Queue <https://docs.python.org/3.5/library/multiprocessing.html#multiprocessing.Queuee>`_
 - организация вычислений из эвентлупа и мультипроцессинга, точно пока не знаю
 - хранилище данных `sqlite3 <https://sqlite.org/index.html>`_
 - шаблонизатор на фронэнде с компиляцией в js `Monkberry <https://monkberry.js.org/>`_
 - `native browser api <https://developer.mozilla.org/en-US/docs/Web/API>`_ для всего остального во фронэнде
 - `ES <https://www.ecma-international.org/publications/standards/Ecma-262.htm>`_ какой-то там

WebSockets не балансируются. Можно было бы браузерный лог полинг,
`Server-sent events <https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events>`_, но
`Edge UNDER CONSIDERATION <https://developer.microsoft.com/en-us/microsoft-edge/platform/status/serversenteventseventsource/?q=EventSource>`_.

Celery - инструмент как раз для таких задач. Но я его ещё не использовал в таком качестве - только как средство выноса
долгих задач из быстрых и коротких обработчиков http
(`django views <https://docs.djangoproject.com/en/2.1/topics/http/views/>`_). Сейчас я не горю особым желанием
разбираться в том, как обращаться с этой очередью для выполнения требований задания. Я расставил приоритеты по другому.
Оставлю celery на другой раз.


План
----

.. role:: strike

1. Web 1.0 версия

   Я не определился с клиентской реализацией, то есть не утвердил варианты ответов на вопросы как, что, зачем. Начну с
   сервера в старом добром стиле сервер-сайд рендеринга и ручным обновлением страницы. Так я продвинусь по заданию,
   выбросив из него `асинхронно`.

   Жаль, но из коробки в aiohttp нет csrf-защиты. Я нашел только `заброшенную работу
   <https://github.com/asvetlov/aiohttp-csrf/tree/init>`_ в отдельном бранче без пакета на pypi.

2. :strike:`Интерактивная версия` Отслеживание процесса вычисления

   Я не увидел слона в посудной лавке. Прежде чем оживлять интерфейс с помощью js, нужно расставить события по ходу
   расчета очередного знака числа пи. Не просто расставить, а спроектировать возможность реагировать на события и
   уведомлять пользователя о прогрессе.

   По хорошему нужно вынести дамп данных в json в отдельный процесс. Операция такая же "дорогая", как и вычисления. Силы
   покидают меня, сделаю задачу-поллинг в цикле веб-сервера.

3. Интерактивная версия

   Я уже не хочу загадывать... но верю, что план на шаге 2 не провалится.

   Реализован ресурс по ws. По нему передаются все актуальные расчеты. Дело за малым: взять данные, передать в шаблон,
   результатом заменить текущие ноды в ДОМе.

4. Страница под конкретную цифру

   Не забыть сделать.


Postmortem
==========

На задание было потрачено 16 h 4 min. Достаточно большой объем времени на тестовое задание из 2 страниц, 2 асинхронных
операций с данными и 2 динамическими элементами на странице. Я не знаю как меня оценят, но я доволен проделанной работе
и полученным результатом. Все что я запланировал мне удалось воплотить в код.


life.js
import Monkberry from './node_modules/monkberry/monkberry.js'
import control from './control.monk'
import table from './table.monk'


let domready = Promise.resolve()
if (document.readyState === 'loading') {
	domready = new Promise(function (resolve, reject) {
		document.addEventListener('DOMContentLoaded', resolve)})}


function onmessage(controlView, tableView, event) {
	const data = JSON.parse(event.data)
	render(controlView, tableView, data)}


function renderTemplate(event) {
	event.target.removeEventListener('message', renderTemplate)

	const $tableContainer = document.getElementById('table')
	while ($tableContainer.firstChild)
		$tableContainer.removeChild($tableContainer.firstChild)
	const tableView = Monkberry.render(table, $tableContainer)

	const $controlContainer = document.getElementById('control')
	while ($controlContainer.firstChild)
		$controlContainer.removeChild($controlContainer.firstChild)
	const controlView = Monkberry.render(control, $controlContainer);

	onmessage(controlView, tableView, event)
	event.target.addEventListener('message', onmessage.bind(null, controlView, tableView))}


function render(controlView, tableView, results) {
	// XXX: https://github.com/antonmedv/monkberry/issues/36
	for (const item of results)
		if (item.result === null)
			item.result = ''
	const next_digit_number = results[0].digit_number + 1
	controlView.update({'next_digit_number': next_digit_number})
	tableView.update({'results': results})}


domready.then(function () {
	// XXX: Last second hack. There is no table for empty database
	if (document.getElementById('table') === null) {
		document.body.removeEventListener('submit', run)
		return}

	const ws = new WebSocket(`ws://${document.location.host}/subscribe`);
	ws.addEventListener('message', renderTemplate)})


function run(event) {
	const target = event.target
	if (!'parentElement' in target || !target.parentElement.matches('#control'))
		return;
	event.preventDefault()
	const request = new Request('/run', {
		method: 'POST',
		headers: new Headers({
			'X-REQUESTED-WITH': 'XMLHttpRequest'}),
		credentials: 'same-origin',
		body: new FormData(target)})
	fetch(request)}

document.body.addEventListener('submit', run)


view.js

/**
 * @class
 */
function control() {
  Monkberry.call(this);

  // Create elements
  var form0 = document.createElement('form');
  var input1 = document.createElement('input');
  var button2 = document.createElement('button');
  var text3 = document.createTextNode('');

  // Construct dom
  input1.setAttribute("type", "hidden");
  input1.setAttribute("name", "next_digit_number");
  button2.appendChild(document.createTextNode("Рассчитать "));
  button2.appendChild(text3);
  button2.appendChild(document.createTextNode(" разряд"));
  button2.setAttribute("class", "btn btn-primary");
  form0.appendChild(input1);
  form0.appendChild(button2);
  form0.setAttribute("class", "form-group");
  form0.setAttribute("method", "post");
  form0.setAttribute("content", "multipart/form-data");

  // Update functions
  this.__update__ = {
    next_digit_number: function (next_digit_number) {
      input1.value = next_digit_number;;
      text3.textContent = next_digit_number;
    }
  };

  // Set root nodes
  this.nodes = [form0];
}
control.prototype = Object.create(Monkberry.prototype);
control.prototype.constructor = control;
control.pool = [];
control.prototype.update = function (__data__) {
  if (__data__.next_digit_number !== undefined) {
    this.__update__.next_digit_number(__data__.next_digit_number);
  }
};

window.control = control;

/**
 * @class
 */
function table() {
  Monkberry.call(this);
  var _this = this;

  // Create elements
  var for0 = document.createComment('for');
  var children0 = new Monkberry.Map();

  // Update functions
  this.__update__ = {
    results: function (results) {
      Monkberry.loop(_this, for0, children0, table_for0, results, {"value":"item"});
    }
  };

  // On update actions
  this.onUpdate = function (__data__) {
    children0.forEach(function (view) {
      view.update(view.__state__);
      view.update(__data__);
      view.update(view.__state__);
    });
  };

  // Set root nodes
  this.nodes = [for0];
}
table.prototype = Object.create(Monkberry.prototype);
table.prototype.constructor = table;
table.pool = [];
table.prototype.update = function (__data__) {
  if (__data__.results !== undefined) {
    this.__update__.results(__data__.results);
  }
  this.onUpdate(__data__);
};

/**
 * @class
 */
function table_for0() {
  Monkberry.call(this);
  this.__state__ = {};
  var _this = this;

  // Create elements
  var tr0 = document.createElement('tr');
  var td1 = document.createElement('td');
  var a2 = document.createElement('a');
  var text3 = document.createTextNode('');
  var td4 = document.createElement('td');
  var for0 = document.createComment('if');
  var child0 = {};
  var for1 = document.createComment('if');
  var child2 = {};
  var for2 = document.createComment('if');
  var child4 = {};

  // Construct dom
  a2.appendChild(text3);
  a2.setAttribute("href", "/detail/");
  td1.appendChild(a2);
  td4.appendChild(for0);
  td4.appendChild(for1);
  td4.appendChild(for2);
  tr0.appendChild(td1);
  tr0.appendChild(td4);

  // Update functions
  this.__update__ = {
    item: function (item) {
      text3.textContent = item.digit_number;
      a2.setAttribute("href", ("/detail/") + (item.digit_number));;
      Monkberry.cond(_this, for0, child0, table_for0_if0, (item.step) !== ('stored'));
      Monkberry.cond(_this, for1, child2, table_for0_if2, ((item.result) !== ('')) && ((item.step) !== ('stored')));
      Monkberry.cond(_this, for2, child4, table_for0_if4, (item.result) !== (''));
    }
  };

  // On update actions
  this.onUpdate = function (__data__) {
    if (child0.ref) {
      child0.ref.update(__data__);
    }
    if (child2.ref) {
      child2.ref.update(__data__);
    }
    if (child4.ref) {
      child4.ref.update(__data__);
    }
  };

  // Set root nodes
  this.nodes = [tr0];
}
table_for0.prototype = Object.create(Monkberry.prototype);
table_for0.prototype.constructor = table_for0;
table_for0.pool = [];
table_for0.prototype.update = function (__data__) {
  if (__data__.item !== undefined && __data__.__index__ !== undefined) {
    this.__update__.item(__data__.item);
  }
  this.onUpdate(__data__);
};

/**
 * @class
 */
function table_for0_if0() {
  Monkberry.call(this);

  // Create elements
  var text0 = document.createTextNode('');

  // Update functions
  this.__update__ = {
    item: function (item) {
      text0.textContent = item.step;
    }
  };

  // Set root nodes
  this.nodes = [text0];
}
table_for0_if0.prototype = Object.create(Monkberry.prototype);
table_for0_if0.prototype.constructor = table_for0_if0;
table_for0_if0.pool = [];
table_for0_if0.prototype.update = function (__data__) {
  if (__data__.item !== undefined) {
    this.__update__.item(__data__.item);
  }
};

/**
 * @class
 */
function table_for0_if2() {
  Monkberry.call(this);

  // Set root nodes
  this.nodes = [document.createTextNode(": ")];
}
table_for0_if2.prototype = Object.create(Monkberry.prototype);
table_for0_if2.prototype.constructor = table_for0_if2;
table_for0_if2.pool = [];
table_for0_if2.prototype.update = function (__data__) {
};

/**
 * @class
 */
function table_for0_if4() {
  Monkberry.call(this);

  // Create elements
  var text0 = document.createTextNode('');

  // Update functions
  this.__update__ = {
    item: function (item) {
      text0.textContent = item.result;
    }
  };

  // Set root nodes
  this.nodes = [text0];
}
table_for0_if4.prototype = Object.create(Monkberry.prototype);
table_for0_if4.prototype.constructor = table_for0_if4;
table_for0_if4.pool = [];
table_for0_if4.prototype.update = function (__data__) {
  if (__data__.item !== undefined) {
    this.__update__.item(__data__.item);
  }
};

window.table = table;
//# sourceMappingURL=view.js.map


table.js
var Monkberry = require('monkberry');

/**
 * @class
 */
function table() {
  Monkberry.call(this);
  var _this = this;

  // Create elements
  var for0 = document.createComment('for');
  var children0 = new Monkberry.Map();

  // Update functions
  this.__update__ = {
    results: function (results) {
      Monkberry.loop(_this, for0, children0, table_for0, results, {"value":"item"});
    }
  };

  // On update actions
  this.onUpdate = function (__data__) {
    children0.forEach(function (view) {
      view.update(view.__state__);
      view.update(__data__);
      view.update(view.__state__);
    });
  };

  // Set root nodes
  this.nodes = [for0];
}
table.prototype = Object.create(Monkberry.prototype);
table.prototype.constructor = table;
table.pool = [];
table.prototype.update = function (__data__) {
  if (__data__.results !== undefined) {
    this.__update__.results(__data__.results);
  }
  this.onUpdate(__data__);
};

/**
 * @class
 */
function table_for0() {
  Monkberry.call(this);
  this.__state__ = {};
  var _this = this;

  // Create elements
  var tr0 = document.createElement('tr');
  var td1 = document.createElement('td');
  var a2 = document.createElement('a');
  var text3 = document.createTextNode('');
  var td4 = document.createElement('td');
  var for0 = document.createComment('if');
  var child0 = {};
  var for1 = document.createComment('if');
  var child2 = {};
  var for2 = document.createComment('if');
  var child4 = {};

  // Construct dom
  a2.appendChild(text3);
  a2.setAttribute("href", "/detail/");
  td1.appendChild(a2);
  td4.appendChild(for0);
  td4.appendChild(for1);
  td4.appendChild(for2);
  tr0.appendChild(td1);
  tr0.appendChild(td4);

  // Update functions
  this.__update__ = {
    item: function (item) {
      text3.textContent = item.digit_number;
      a2.setAttribute("href", ("/detail/") + (item.digit_number));;
      Monkberry.cond(_this, for0, child0, table_for0_if0, (item.step) !== ('stored'));
      Monkberry.cond(_this, for1, child2, table_for0_if2, ((item.result) !== ('')) && ((item.step) !== ('stored')));
      Monkberry.cond(_this, for2, child4, table_for0_if4, (item.result) !== (''));
    }
  };

  // On update actions
  this.onUpdate = function (__data__) {
    if (child0.ref) {
      child0.ref.update(__data__);
    }
    if (child2.ref) {
      child2.ref.update(__data__);
    }
    if (child4.ref) {
      child4.ref.update(__data__);
    }
  };

  // Set root nodes
  this.nodes = [tr0];
}
table_for0.prototype = Object.create(Monkberry.prototype);
table_for0.prototype.constructor = table_for0;
table_for0.pool = [];
table_for0.prototype.update = function (__data__) {
  if (__data__.item !== undefined && __data__.__index__ !== undefined) {
    this.__update__.item(__data__.item);
  }
  this.onUpdate(__data__);
};

/**
 * @class
 */
function table_for0_if0() {
  Monkberry.call(this);

  // Create elements
  var text0 = document.createTextNode('');

  // Update functions
  this.__update__ = {
    item: function (item) {
      text0.textContent = item.step;
    }
  };

  // Set root nodes
  this.nodes = [text0];
}
table_for0_if0.prototype = Object.create(Monkberry.prototype);
table_for0_if0.prototype.constructor = table_for0_if0;
table_for0_if0.pool = [];
table_for0_if0.prototype.update = function (__data__) {
  if (__data__.item !== undefined) {
    this.__update__.item(__data__.item);
  }
};

/**
 * @class
 */
function table_for0_if2() {
  Monkberry.call(this);

  // Set root nodes
  this.nodes = [document.createTextNode(": ")];
}
table_for0_if2.prototype = Object.create(Monkberry.prototype);
table_for0_if2.prototype.constructor = table_for0_if2;
table_for0_if2.pool = [];
table_for0_if2.prototype.update = function (__data__) {
};

/**
 * @class
 */
function table_for0_if4() {
  Monkberry.call(this);

  // Create elements
  var text0 = document.createTextNode('');

  // Update functions
  this.__update__ = {
    item: function (item) {
      text0.textContent = item.result;
    }
  };

  // Set root nodes
  this.nodes = [text0];
}
table_for0_if4.prototype = Object.create(Monkberry.prototype);
table_for0_if4.prototype.constructor = table_for0_if4;
table_for0_if4.pool = [];
table_for0_if4.prototype.update = function (__data__) {
  if (__data__.item !== undefined) {
    this.__update__.item(__data__.item);
  }
};

module.exports = table;
//# sourceMappingURL=table.js.map


server.py
import asyncio
import json
import logging
import multiprocessing
import weakref
from asyncio import sleep
from decimal import Decimal, getcontext
from enum import Enum
from itertools import chain
from math import factorial
from multiprocessing import Process, Queue, cpu_count
from pathlib import Path

import aiohttp_jinja2
import jinja2
from aiohttp import web, WSCloseCode
from functools import partial
from sqlalchemy import create_engine, event, Column, types, UniqueConstraint, Index, ForeignKey, func
from sqlalchemy.engine import Engine
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, relationship

# XXX: Singletons. Singletons everywhere!

ROOT_PATH = Path(__file__).parent


@event.listens_for(Engine, 'connect')
def set_sqlite_pragma(dbapi_connection, connection_record):
	cursor = dbapi_connection.cursor()
	cursor.execute('PRAGMA journal_mode=WAL')
	cursor.close()


Base = declarative_base()


class DigitNumber(Base):
	__tablename__ = 'digit_number'

	digit_number = Column(types.Integer, primary_key=True)

	def __repr__(self):
		return 'DigitNumber(digit_number={self.digit_number!r})'.format(self=self)


Progress = Enum('Progress', 'registered scheduled taken calculated stored', module=__name__)


class Event(Base):
	__tablename__ = 'event'

	id = Column(types.Integer, primary_key=True)
	digit_number = Column(types.Integer, ForeignKey(DigitNumber.digit_number))
	progress = Column(types.SmallInteger)
	result = Column(types.String, default=None)

	__table_args__ = UniqueConstraint('digit_number', 'progress'),


# XXX: Copy-pasted from http://blog.recursiveprocess.com/2013/03/14/calculate-pi-with-python/
def chudnovsky(n): #http://en.wikipedia.org/wiki/Chudnovsky_algorithm
	getcontext().prec = n + 100
	pi = Decimal(0)
	k = 0
	while k < n:
		pi += (Decimal(-1)**k)*(Decimal(factorial(6*k))/((factorial(k)**3)*(factorial(3*k)))*(13591409+545140134*k)/(640320**(3*k)))
		k += 1
	pi = pi * Decimal(10005).sqrt()/4270934400
	pi = pi**(-1)
	return pi


def worker(task_queue, engine, database_updated):
	Session = sessionmaker()
	Session.configure(bind=engine)
	session = Session()

	while True:
		try:
			digit_number = task_queue.get()
		except (Exception, KeyboardInterrupt):
			break

		session.add(Event(
			digit_number=digit_number,
			progress=Progress.taken.value))
		try:
			session.commit()
			database_updated.set()
		except IntegrityError:
			session.rollback()
			continue

		try:
			template = '{{:.{:d}}}'.format(digit_number+1)
			number = template.format(chudnovsky(digit_number))
			calculated_error = None
		except Exception as e:
			calculated_error = repr(e)

		session.add(Event(
			digit_number=digit_number,
			progress=Progress.calculated.value,
			result=calculated_error))
		session.commit()
		database_updated.set()

		if calculated_error is None:
			session.add(Event(
				digit_number=digit_number,
				progress=Progress.stored.value,
				result=number))
			session.commit()
			database_updated.set()


async def publisher(database_updated, websockets, engine):
	Session = sessionmaker()
	Session.configure(bind=engine)
	session = Session()

	while True:
		if not database_updated.is_set():
			await sleep(16/1000)
			continue

		database_updated.clear()
		results_query = (session
			.query(DigitNumber.digit_number, Event.progress, Event.result)
			.join(Event)
			.group_by(DigitNumber)
			.having(Event.progress == func.MAX(Event.progress))
			.order_by(-DigitNumber.digit_number))

		results = []
		for digit_number, step, result in results_query:
			results.append({
				'digit_number': digit_number,
				'step': Progress(step).name,
				'result': result})
		dump = json.dumps(results)

		for ws in set(websockets):
			await ws.send_str(dump)


routes = web.RouteTableDef()
routes.static('/bootstrap', ROOT_PATH/'node_modules'/'bootstrap'/'dist', name='bootstrap')
routes.static('/monkberry', ROOT_PATH/'node_modules'/'monkberry', name='monkberry')
routes.static('/static', ROOT_PATH, name='self')


@aiohttp_jinja2.template('index.html')
async def index(request):
	session = request.app['session']
	database_updated = request.app['database_updated']
	results = (session
		.query(DigitNumber.digit_number, Event.progress, Event.result)
		.join(Event)
		.group_by(DigitNumber)
		.having(Event.progress == func.MAX(Event.progress))
		.order_by(-DigitNumber.digit_number))
	results_exists = session.query(results.exists()).scalar()
	next_digit_number = 1
	if results_exists:
		results = iter(results)
		first = next(results)
		next_digit_number = first[0] + 1
		results = chain((first,), results)

	if request.method == 'POST':
		data = await request.post()
		request_digit_number = None
		try:
			request_digit_number = int(data['next_digit_number'])
		except ValueError:
			pass

		if request_digit_number is not None and request_digit_number >= next_digit_number:
			session.add(DigitNumber(digit_number=request_digit_number))
			session.add(Event(
				digit_number=request_digit_number,
				progress=Progress.registered.value))

			try:
				session.commit()
				database_updated.set()
			except IntegrityError:
				session.rollback()

			else:
				try:
					request.app['task_queue'].put(request_digit_number, timeout=1)
					scheduled_error = None
				except Exception as e:
					scheduled_error = repr(e)

				session.add(Event(
					digit_number=request_digit_number,
					progress=Progress.scheduled.value,
					result=scheduled_error))
				session.commit()
				database_updated.set()

		raise web.HTTPFound(request.path)

	return {
		'next_digit_number': next_digit_number,
		'results_exists': results_exists,
		'results': results,
		'Progress': Progress}


@routes.post('/run')
async def run(request):
	session = request.app['session']
	database_updated = request.app['database_updated']
	results = (session
		.query(DigitNumber.digit_number)
		.order_by(-DigitNumber.digit_number)
		[:1])
	next_digit_number = 1
	if results:
		next_digit_number = results[0][0] + 1

	data = await request.post()
	request_digit_number = None
	try:
		request_digit_number = int(data['next_digit_number'])
	except ValueError:
		pass

	if request_digit_number is not None and request_digit_number >= next_digit_number:
		session.add(DigitNumber(digit_number=request_digit_number))
		session.add(Event(
			digit_number=request_digit_number,
			progress=Progress.registered.value))

		try:
			session.commit()
			database_updated.set()
		except IntegrityError:
			session.rollback()

		else:
			try:
				request.app['task_queue'].put(request_digit_number, timeout=1)
				scheduled_error = None
			except Exception as e:
				scheduled_error = repr(e)

			session.add(Event(
				digit_number=request_digit_number,
				progress=Progress.scheduled.value,
				result=scheduled_error))
			session.commit()
			database_updated.set()

	return web.Response(text='OK')


@routes.get('/detail/{digit_number}', name='detail')
@aiohttp_jinja2.template('detail.html')
async def detail(request):
	session = request.app['session']
	results = (session
		.query(DigitNumber.digit_number, Event.progress, Event.result)
		.join(Event)
		.group_by(DigitNumber)
		.having(Event.progress == func.MAX(Event.progress))
		.order_by(-DigitNumber.digit_number)
		.filter(DigitNumber.digit_number==request.match_info['digit_number'])
		[:1])

	if not results:
		raise web.HTTPNotFound()

	return {
		'digit_number': results[0][0],
		'step': results[0][1],
		'result': results[0][2],
		'Progress': Progress}


async def subscribe(request):
	ws = web.WebSocketResponse()
	await ws.prepare(request)

	request.app['websockets'].add(ws)
	try:
		async for _ in ws:
			pass
	finally:
		request.app['websockets'].discard(ws)

	return ws


async def on_shutdown(app):
	for ws in set(app['websockets']):
		await ws.close(
			code=WSCloseCode.GOING_AWAY)


def main():
	logging.basicConfig(level=logging.DEBUG)

	engine = create_engine('sqlite:///{!s}'.format(ROOT_PATH/'db.sqlite3'))
	Base.metadata.create_all(engine)

	Session = sessionmaker()
	Session.configure(bind=engine)

	task_queue = Queue()
	database_updated = multiprocessing.Event()
	workers = []
	for _ in range(cpu_count()):
		background = Process(target=worker, args=(task_queue, engine, database_updated), daemon=True)
		background.start()
		workers.append(background)

	websockets = weakref.WeakSet()
	loop = asyncio.get_event_loop()
	publisher_task = loop.create_task(publisher(database_updated, websockets, engine))

	app = web.Application()
	app.add_routes(routes)
	app.router.add_route('*', '/', index, name='index')
	app.add_routes([web.get('/subscribe', subscribe)])
	app['session'] = Session()
	app['task_queue'] = task_queue
	app['database_updated'] = database_updated
	app['websockets'] = websockets
	app.on_shutdown.append(on_shutdown)
	#app.on_shutdown.append(async lambda app: publisher_task.cancel()) #XXX: How to cancel task?

	with (ROOT_PATH / 'index.html').open('rt') as index_html, (ROOT_PATH / 'detail.html').open('rt') as detail_html:
		aiohttp_jinja2.setup(
			app,
			loader=jinja2.DictLoader({
				'index.html': index_html.read(),
				'detail.html': detail_html.read()}))

	try:
		web.run_app(app, port=50594)

	finally:
		task_queue.close()
		for background in workers:
			background.terminate()
		for background in workers:
			background.join()
		task_queue.join_thread()


if __name__ == '__main__':
	main()
